# RNA-seq Alignent

Welcome to an introduction of RNA-seq analysis, a part of the High Throughput Sequencing Facility introduction to Deep Sequencing Analysis.

## 0. Setup

### Log in

Login to Killdevil with your onyen.

### Load modules

One of the first things we'll have to do is load up all our modules. On Kure and Killdevil, there are many applications and programs pre-installed. To use them, you simply need to load them!

The applications we'll load up for today will be:  
fastx_toolkit   
tophat2  
htseq-count
r    
samtools   
bedtools
bamtools


To load them, type:

```
$ module avail            # List all the modules available.

$ module load tophat     # Load tophat2
$ module load htseq-count #Load htseq-count
$ module load r           # Load r
$ module load samtools    # Load samtools
#$ module load bedtools    # Load bedtools
#$ module load bamtools    # Load bamtools

$ module list             # List all loaded modules
```

To learn more about modules, read [Getting Started with Killdevil -- Application Environment](http://help.unc.edu/help/getting-started-on-killdevil/#P92_10973).


### Start a directory structure

It is REALLY important to stay organized while performing computational projects. If you can't keep track or keep a record of what you are doing, you will have to do it again... or worse, you might make a mistake! So be efficient by taking good notes and keeping things organized.

Let's make a file structure. 

Start by creating the directory where you will conduct your analysis. I typically give my project ProjectNumbers.

```
$ mkdir 01_RNASeqDemo
$ cd 01_RNASeqDemo
```

OK, now we are inside the directory `01_RNASeqDemo`. Now that we are here, initiate the project by starting some sub-directories and a README document.

```
mkdir 00_logs
mkdir 01_ref
mkdir 02_raw
mkdir 03_processedInput
mkdir 04_resultsInput
mkdir 05_scripts
mkdir 06_test
touch 01_RNASeqDemo_README.txt
```


### Get the dataset:

We will obtain a demonstration dataset. We are going to use a set of 10 files that came from 10 different RNAseq libraries (5
Male cell lines, 5 female cell lines).

Acknowledgement: This data was generated by Dr. Mauro Calabrese (Assistant Professor, Pharmacology, UNC-CH). We are using a small subset of reads in interest of time. A real dataset will be multiple times this size. You will find the data files in /proj/seq/data/RNAseq-HSL/reads/ directory.   
We have two groups of cell-lines. Samples are divided into five male and five female cell lines. These are paired-end 75 bp Illumina NextSeq 500 reads.

Gm10847 F   
Gm12753 F   
Gm12802 F   
Gm12818 F   
Gm12832 F   

Gm10851 M   
Gm12752 M   
Gm12801 M   
Gm12864 M   
Gm12877 M   

Copy these files to the directory called /raw within your project directory:


```
cd 02_rawInput
cp /proj/seq/data/RNAseq-HSL/reads/*.fastq.gz .
ls
```

### Get the scripts

I have pre-written a number of scripts for you to run as examples. To obtain these, navigate to `~/01_RNASeqDemo/05_scripts/` and copy the following files like so...

```
pwd                                                     # Should show that you are in your version of ~/01_RNASeqDemo/05_scripts/
cp /netscr/erinosb/HTSF_RNASeq_Demo/05_scripts/*.sh .   # Copy scripts from erinosb location to your local 05_scripts area
```





## The basics of a tophat command

To run tophat, simply execute...

> $ **tophat**&emsp;*[options]*&emsp;*\<genome_index_base\>&emsp;\<reads1_1[,...,readsN_1]\>&emsp;[reads1_2,...readsN_2]*

```
tophat                      # the literal command
[options]                   # any optional arguments to customize the way that tophat runs
<genome_index_base>         # The genome you want to reference, built as a bowtie2 index and written in the form /path/to/rootname
<reads1_1[,...,readsN_1]    # A list of comma-separated fastq files to align
[reads1_2,...readsN_2]      # If you are doing paired-end sequencing, the matching 'right-hand' sequences you want to align
```

Running tophat on our first sample...

```
tophat
  -p 4                                                       # use four threads
  --max-multihits 1                                          # don't report any reads with more than one match to the genome
  -o ../04_results/tophat/Gm10847_opd                        # output directory
  -G /proj/seq/data/HG19_UCSC/Annotation/Genes/genes.gtf     # Use a transcriptome first
  /proj/seq/data/HG19_UCSC/Sequence/Bowtie2Index/genome      # This is the reference genome index (Bowtie2)
  ../03_processedInput/Gm10847_R1_trim.fastq.gz              # Input .fastq reads to align... paired-end R1
  ../03_processedInput/Gm10847_R2_trim.fastq.gz              # Input .fastq reads to align... paired end R2
```

To execute the tophat command, we need to use the Load Sharing Facility (LSF) using bsub commands.

ONE OPTION... 

We could execute the command within a bsub command...

```
bsub -q week -n 4 -R "span[hosts=1]" -o %J_tophat.log "tophat -p 4 --max-multihits 1 -o ../04_results/tophat/Gm10847_opd -G /proj/seq/data/HG19_UCSC/Annotation/Genes/genes.gtf /proj/seq/data/HG19_UCSC/Sequence/Bowtie2Index/genome ../03_processedInput/Gm10847_R1_trim.fastq.gz ../03_processedInput/Gm10847_R2_trim.fastq.gz"

NOTE: -p in the tophat command and -n in the bsub command must be compatible
```

Wow! This is getting really cluttered fast! It is often easier to package the tophat commands into a shell script for ease of execution.

In this case, the script will look like...

**script01_trim.sh**
```
#!/bin/bash

tophat -p 4 --max-multihits 1 -o ../04_results/tophat/Gm10847_opd -G /proj/seq/data/HG19_UCSC/Annotation/Genes/genes.gtf /proj/seq/data/HG19_UCSC/Sequence/Bowtie2Index/genome ../03_processedInput/Gm10847_R1_trim.fastq.gz ../03_processedInput/Gm10847_R2_trim.fastq.gz
```

Then you can execute the command like so...

> bsub -q week -n 4 -R "span[hosts=1]" -o %J_tophat.log "bash script01_trim.sh"











