# RNA-seq Alignent

Welcome to an introduction of RNA-seq analysis, a part of the High Throughput Sequencing Facility introduction to Deep Sequencing Analysis.

## 0. Setup

### Log in

Login to Killdevil with your onyen.

### Load modules

One of the first things we'll have to do is load up all our modules. On Kure and Killdevil, there are many applications and programs pre-installed. To use them, you simply need to load them!

The applications we'll load up for today will be:  
fastx_toolkit   
tophat2  
htseq-count
r    
samtools   
bedtools   


To load them, type:

```
$ module avail            # List all the modules available.

$ module load tophat2     # Load tophat2
$ module load htseq-count #Load htseq-count
$ module load r           # Load r
$ module load samtools    # Load samtools
$ module load bedtools    # Load bedtools

$ module list             # List all loaded modules
```

To learn more about modules, read [Getting Started with Killdevil -- Application Environment](http://help.unc.edu/help/getting-started-on-killdevil/#P92_10973).


### Start a directory structure

It is REALLY important to stay organized while performing computational projects. If you can't keep track or keep a record of what you are doing, you will have to do it again... or worse, you might make a mistake! So be efficient by taking good notes and keeping things organized.

Let's make a file structure. 

Start by creating the directory where you will conduct your analysis. I typically give my project ProjectNumbers.

```
$ mkdir 01_RNASeqDemo
$ cd 01_RNASeqDemo
```

OK, now we are inside the directory `01_RNASeqDemo`. Now that we are here, initiate the project by starting some sub-directories and a README document.

```
$ mkdir 00_logs
$ mkdir 01_ref
$ mkdir 02_raw
$ mkdir 03_processed
$ mkdir 04_results
$ mkdir 05_bin
$ mkdir 06_test
$ touch 01_RNASeqDemo_README.txt
```

### Get the dataset:

We will obtain a demonstration dataset. We are going to use a set of 10 files that came from 10 different RNAseq libraries (5
Male cell lines, 5 female cell lines).

Acknowledgement: This data was generated by Dr. Mauro Calabrese (Assistant Professor, Pharmacology, UNC-CH). We are using a small subset of reads in interest of time. A real dataset will be multiple times this size. You will find the data files in /proj/seq/data/RNAseq-HSL/reads/ directory.   
We have two groups of cell-lines. Samples are divided into five male and five female cell lines. These are paired-end 75 bp Illumina NextSeq 500 reads.

Gm10847 F   
Gm12753 F   
Gm12802 F   
Gm12818 F   
Gm12832 F   

Gm10851 M   
Gm12752 M   
Gm12801 M   
Gm12864 M   
Gm12877 M   

Copy these files to the directory called /raw within your project directory:


```
$ cd 02_raw
$ cp /proj/seq/data/RNAseq-HSL/reads/*.fastq.gz .
$ ls
```



