# RNA-seq Alignent

Welcome to an introduction of RNA-seq analysis, a part of the High Throughput Sequencing Facility introduction to Deep Sequencing Analysis.

&nbsp;

&nbsp;

____

## Before we get started...

+ \* -- wild card
+ # -- comment out
+ Relative paths versus absolute paths
+ On command line notation
+ On /proj/data/seq
+ On annotation files


&nbsp;

&nbsp;

____

## Setup

#### Log in

Login to Killdevil with your onyen.

#### Load modules

One of the first things we'll have to do is load up all our modules. On Kure and Killdevil, there are many applications and programs pre-installed. To use them, you simply need to load them!

The applications we'll load up for today will be:  
fastx_toolkit   
tophat2  
bedtools   
bamtools  
pico


To load them, type:

```
#EXECUTE THESE
$ module avail             # List all the modules available.

$ module load tophat       # Load tophat2
$ module load samtools     # Load samtools
$ module load bedtools     # Load bedtools
$ module load bamtools     # Load bamtools
$ module load pico         # Load pico
$ module list              # List all loaded modules
```

To learn more about modules, read [Getting Started with Killdevil -- Application Environment](http://help.unc.edu/help/getting-started-on-killdevil/#P92_10973).


#### Start a directory structure

It is REALLY important to stay organized while performing computational projects. If you can't keep track of what you have done, you will have to do it again... or worse, you might make a mistake! 

So be efficient by taking good notes and keeping things organized.

Start by creating the directory where you will conduct your analysis. 

```
#NAVIGATE TO your prefered working directory. Either /netscr/<onyen>/ or /proj/<yourlab>/<yourname>/ We will call this location <yourversion>

#EXECUTE THESE COMMANDS
$ pwd                             # Use this to check that you are located where you want to be. It is probably /netscr/<onyen>/
$ mkdir 01_RNASeqDemo             # Use this to start a directory where we will work on our RNA-seq project today.
$ cd 01_RNASeqDemo                # Change into that directory
```

OK, now we are inside the directory `01_RNASeqDemo`. Now that we are here, initiate the project by starting some sub-directories and a README document.

```
#NAVIGATE TO <yourversion>/01_RNASeqDemo/
$ pwd                             # Use this to check that you are located where you want to be. It is probably /netscr/<onyen>/01_RNASeqDemo/ 

#EXECUTE THESE COMMANDS
$ mkdir 00_logs 01_ref 02_rawInput 03_processedInput 04_results 05_scripts 06_test
$ touch RNASeqDemo_README.txt
$ ls
```


#### Get the dataset:

We will obtain a demonstration dataset. We are going to use a set of 10 files that came from 10 different RNAseq libraries (5
Male cell lines, 5 female cell lines).

Acknowledgement: This data was generated by Dr. Mauro Calabrese (Assistant Professor, Pharmacology, UNC-CH). We are using a small subset of reads in interest of time. A real dataset will be multiple times this size. You will find the data files in `/proj/seq/data/RNAseq-HSL/reads/` directory.  

We have two groups of cell-lines. Samples are divided into five male and five female cell lines. These are paired-end 75 bp Illumina NextSeq 500 reads.

Gm10847 F   
Gm12753 F   
Gm12802 F   
Gm12818 F   
Gm12832 F   

Gm10851 M   
Gm12752 M   
Gm12801 M   
Gm12864 M   
Gm12877 M   

Copy these files to the directory called `02_rawInput` within your project directory:

```
#NAVIGATE TO <yourversion>/01_RNASeqDemo/02_rawInput
$ cd 02_rawInput
#EXECUTE THESE COMMANDS:
$ pwd                                               # Should show you <yourversion>/01_RNASeqDemo/02_rawInput
$ cp /proj/seq/data/RNAseq-HSL/reads/*.fastq.gz .   # Copy a dataset from the HTSF shared data directory into your 02_rawInput directory.
$ ls                                                # Look and see what you got!
```

We don't have time to do the trimming of these files so I've trimmed the files for you. Download the trimmed files.

```
#NAVIGATE TO <yourversion>/01_RNASeqDemo/03_processedInput
$ cd ../03_processedInput
#EXECUTE THESE COMMANDS:
$ pwd                                               # Should show you <yourversion>/01_RNASeqDemo/03_processedInput
$ cp /netscr/erinosb/HTSF_RNASeq_Demo/01_RNASeqDemo/03_processedInput/*.fastq.gz .   # Copy a dataset from the HTSF shared data directory into your 02_rawInput directory.
$ ls 
```


#### Get the scripts

I have pre-written a number of scripts for you to run as examples. To obtain these, navigate to `<yourversion>/01_RNASeqDemo/05_scripts/` and copy the following files like so...

```
#NAVIGATE TO <yourversion>/01_RNASeqDemo/05_scripts/
$ cd ../05_scripts
#EXECUTE THESE COMMANDS:
$ pwd                                     # Should show that you are in <yourversion>/01_RNASeqDemo/05_scripts/
$ cp /netscr/erinosb/HTSF_RNASeq_Demo/01_RNASeqDemo/05_scripts/*.sh .   # Copy scripts from erinosb location to your local 05_scripts area
$ ls                                      # Look at what you have
```


&nbsp;

&nbsp;

____
## Align your first sample using tophat

We are going to align one sample using tophat. It should take about 20 minutes and will be done by the time we are finished with the introductory lecture. To execute this command, follow these directions

```
#NAVIGATE TO <yourversion>/01_RNASeqDemo/05_scripts/
#EXECUTE THESE COMMANDS:
$ pwd                                             # Should show that you are in <yourversion>/01_RNASeqDemo/05_scripts/
$ mkdir ../04_results/tophat                      # Make an output directory to capture all tophat output
$ mkdir ../04_results/tophat/Gm10847_opd          # Save an output directory to capture your output for the first sample
$ ls ../04_results/tophat                         # Make sure your directory making worked

### HERE YOU GO!!! YOU"RE GOING TO START TOPHAT ###
#EXECUTE THIS COMMAND:
$ pwd                                             # Should show that you are in <yourversion>/01_RNASeqDemo/05_scripts/
$ bsub -q week -n 4 -R "span[hosts=1]" -o %J_tophat.log "bash script02_tophat_single.sh"     #Start the tophat demo.
```

___________
## STOP HERE
________
&nbsp;

&nbsp;
_______

## The basics of a tophat command

To run tophat, simply execute...

> $ **tophat**&emsp;*[options]*&emsp;*\<genome_index_base\>&emsp;\<reads1_1[,...,readsN_1]\>&emsp;[reads1_2,...readsN_2]*

```
################ DON'T EXECUTE #########################
################ EXAMPLE ONLY ##########################
tophat                      # the literal command
[options]                   # any optional arguments to customize the way that tophat runs
<genome_index_base>         # The genome you want to reference, built as a bowtie2 index and written in the form /path/to/rootname
<reads1_1[,...,readsN_1]    # A list of comma-separated fastq files to align
[reads1_2,...readsN_2]      # If you are doing paired-end sequencing, the matching 'right-hand' sequences you want to align
```

#### Our parameters for running tophat on the first of 10 samples...

Tophat has a lot of options. These allow us to customize how tophat runs based on our personal preferences. For a full list of tophat options, consult the [Tophat manual](https://ccb.jhu.edu/software/tophat/manual.shtml).

Today we will run tophat on our first of ten sequence samples like so...

```
################ DON'T EXECUTE #########################
################ EXAMPLE ONLY ##########################

tophat
  -p 4                                                       # use four threads
  --max-multihits 1                                          # don't report any reads with more than one match to the genome
  -o ../04_results/tophat/Gm10847_opd                        # output directory
  -G /proj/seq/data/HG19_UCSC/Annotation/Genes/genes.gtf     # Use a transcriptome first
  /proj/seq/data/HG19_UCSC/Sequence/Bowtie2Index/genome      # This is the reference genome index (Bowtie2)
  ../03_processedInput/Gm10847_R1_trim.fastq.gz              # This is the input .fastq file to be aligned... paired-end R1
  ../03_processedInput/Gm10847_R2_trim.fastq.gz              # This is the input .fastq file to be aligned... paired end R2
```

To execute this command, we will use the Load Sharing Facility (LSF) using bsub commands. There are a few main ways to do this.

ONE OPTION... 

We *could* execute the command within a bsub command like so:

```
################ DON'T EXECUTE #########################
################ EXAMPLE ONLY ##########################

bsub -q week -n 4 -R "span[hosts=1]" -o ../00_logs/%J_tophat.log "tophat -p 4 --max-multihits 1 -o ../04_results/tophat/Gm10847_opd -G /proj/seq/data/HG19_UCSC/Annotation/Genes/genes.gtf /proj/seq/data/HG19_UCSC/Sequence/Bowtie2Index/genome ../03_processedInput/Gm10847_R1_trim.fastq.gz ../03_processedInput/Gm10847_R2_trim.fastq.gz"

NOTE: -p in the tophat command and -n in the bsub command must be compatible
```

Wow! This is getting really cluttered fast!  

It is often easier to package the tophat commands into a shell script for ease of execution. Shell scripts are files containing a list of commands you want to be executed. In this case, our very basic shell script will read as below.

**script02_tophat_single.sh**
```bash
################ DON'T EXECUTE #########################
################ EXAMPLE ONLY ##########################

#!/bin/bash

tophat -p 4 --max-multihits 1 -o ../04_results/tophat/Gm10847_opd -G /proj/seq/data/HG19_UCSC/Annotation/Genes/genes.gtf /proj/seq/data/HG19_UCSC/Sequence/Bowtie2Index/genome ../03_processedInput/Gm10847_R1_trim.fastq.gz ../03_processedInput/Gm10847_R2_trim.fastq.gz
```

To execute the script, you already typed the following...

```
################ DON'T EXECUTE #########################    
################ EXAMPLE ONLY ##########################    
$ bsub -q week -n 4 -R "span[hosts=1]" -o %J_tophat.log "bash script02_tophat_single.sh"
```

Let's take a closer look at bash script02\_tophat\_single.sh

```
#NAVIGATE TO <yourversion>/01_RNASeqDemo/05_scripts/.
#EXECUTE THESE COMMANDS:
$ more script02_tophat_single.sh
```

&nbsp;

&nbsp;

____
## Tophat output

#### Output files -- you made a log file. 

Let's see if it worked. You wrote an output file as part of your busb command execution. Navigate to `<yourversion>/01_RNASeqDemo/05_scripts/`. 

Use `ls` to find your log file. 

Read the output.  

#### Tophat Output Files  

Now navigate to the location where tophat dumped all its results.

Since we specified to put our output files in `<yourlocation>/01_RNASeqDemo/04_results/tophat`. Go to that location and explore the output. Tophat will produce 8 output files/directories. 

> Navigate to \<yourlocation\>/01_RNASeqDemo/04_results/tophat

**accepted_hits.bam**   
This is the main alignment output file! This is a list of read alignments in BAM format (bam is the binary version).
If you want to convert the .bam file to something you can browse, use `samtools view -h -o accepted_hits.sam accepted_hits.bam`

**align_summary.txt**   
Provide alignment summary. 

**insertions.bed/deletions.bed**   
The USCS BED tracks of insertions and deletions reported by TopHat.

**junctions.bed**   
The UCSC BED track of junctions reported by TopHat. 

**logs**   
The log provides information on the completion of each step in TopHat.

**prep_reads.info**   
Provides numbers of reads processed and the minimum/maximum read length in the data set.

**unmapped.bam**   
This one .bam file contains the unaligned reads.

&nbsp;

&nbsp;
____

## Running all samples

Our example only aligned one of or our ten datasets. To run all 10, we can make a script that simply lists all the commands to generate the full aligment dataset:

**script03A_tophat_unlooped.sh**
```bash
################ DON'T EXECUTE #########################
################ EXAMPLE ONLY ##########################
mkdir ../04_results/tophat
mkdir ../04_results/tophat/Gm10847_opd
tophat -p 4 --max-multihits 1 -o ../04_results/tophat/Gm10847_opd -G /proj/seq/data/HG19_UCSC/Annotation/Genes/genes.gtf /proj/seq/data/HG19_UCSC/Sequence/Bowtie2Index/genome ../03_processedInput/Gm10847_R1_trim.fastq.gz ../03_processedInput/Gm10847_R2_trim.fastq.gz

mkdir ../04_results/tophat/Gm10851_opd
tophat -p 4 --max-multihits 1 -o ../04_results/tophat/Gm10851_opd -G /proj/seq/data/HG19_UCSC/Annotation/Genes/genes.gtf /proj/seq/data/HG19_UCSC/Sequence/Bowtie2Index/genome ../03_processedInput/Gm10851_R1_trim.fastq.gz ../03_processedInput/Gm10851_R2_trim.fastq.gz

mkdir ../04_results/tophat/Gm12752_opd
tophat -p 4 --max-multihits 1 -o ../04_results/tophat/Gm12752_opd -G /proj/seq/data/HG19_UCSC/Annotation/Genes/genes.gtf /proj/seq/data/HG19_UCSC/Sequence/Bowtie2Index/genome ../03_processedInput/Gm12752_R1_trim.fastq.gz ../03_processedInput/Gm12752_R2_trim.fastq.gz

mkdir ../04_results/tophat/Gm12753_opd
tophat -p 4 --max-multihits 1 -o ../04_results/tophat/Gm12753_opd -G /proj/seq/data/HG19_UCSC/Annotation/Genes/genes.gtf /proj/seq/data/HG19_UCSC/Sequence/Bowtie2Index/genome ../03_processedInput/Gm12753_R1_trim.fastq.gz ../03_processedInput/Gm12753_R2_trim.fastq.gz

mkdir ../04_results/tophat/Gm12801_opd
tophat -p 4 --max-multihits 1 -o ../04_results/tophat/Gm12801_opd -G /proj/seq/data/HG19_UCSC/Annotation/Genes/genes.gtf /proj/seq/data/HG19_UCSC/Sequence/Bowtie2Index/genome ../03_processedInput/Gm12801_R1_trim.fastq.gz ../03_processedInput/Gm12801_R2_trim.fastq.gz
#and on and on and on and on....

```

However, I find it easier, and less typing, to write a short shell script that loops through every sample automatically.

Look at the script **script03B_tophat_looped.sh** as an example. This script executes in about 5 hours when run with the bsub command listed under USAGE.

This script is written in bash so it is a bash script. If you want to learn more about bash scripts and writing them, I recommend [Ryan's Tutorial](http://ryanstutorials.net/bash-scripting-tutorial/). Another alternative is to learn pretty much any high-level programming language (perl, python, ruby). Once you learn one, the others are easier! They will ALL make these types of looped analysis "pipelines".


&nbsp;

&nbsp;
_______

## Tophat can be made more efficient.

When tophat aligns to a transcriptome first, it does two things:
1. Goes through the .gtf/gff file
2. builds a transcriptome index off of the .gtf/.gff file
3. aligns the reads to the transcriptome index and genome index given.

If we are running multiple samples, we don't need to build the transcriptome index every time. Go to the script **script04_tophat_alternates.sh** to see how you can write a script that uses this more efficient style.

&nbsp;

&nbsp;
____

## **Homework**

+ Read through the pseudocode (English) parts of the script **script04_tophat_alternates.sh**. Try to follow the logic of what is happening.

+ Run the code:

```
#NAVIGATE TO <yourversion>/01_RNASeqDemo/05_scripts/
#EXECUTE THESE COMMANDS:
$ pwd
$ bsub -q week -n 4 -R "span[hosts=1]" -o ../00_logs/%J_tophat_alt.log "bash script04_tophat_alternates.sh"
```

+ Inspect the log file. Where did the log file go? It should be in `<yourversion>/01_RNASeqDemo/00_logs/`
  + How long did it take to run?
+ Inspect the output tophat in `<yourversion>/01_RNASeqDemo/04_results/tophat/

&nbsp;

&nbsp;
____

## Options to modify

Tophat is really flexible! It has a lot of options! Depending on your experiment, you may want to change some of the default settings to something that is more efficient, works for your specific organism, or annotation version.

**Useful flags to adjust**
+ **-o/--output-dir <string>** The output directory. Change this for every sample.
+ **-i/--min-intron-length** The minimum intron length. Default is 70. This is optimized for mammals. For model organisms, you might need to set this smaller.
+ **-I/--max-intron-length** the maximum intron length. Deault is 500000. This is optimized for mammals. For model organisms, you might need to set this smaller.
+ **-r/--mate-inner-dist** Expected (mean) inner distance between mate pairs for paired end reads. Default is 50bp.
+ **-N/--read-mismatches** Find read alignments having more than these many mismatches are discarded. Default is 2.
+ **--g/--max-multihits <int>** Sets how many possible locations in the genome a given read can match to and still be retained as a successful alignment.
+ **bowtie2 reference files** Change this for each organism you use. There are many organism's .bt2 files already preloaded on the cluster for you at `/proj/seq/data/`. If you need to build your own bt2 files, instructions are here:
+ **.fastq.gz** Of course, this will change for each sample!























